File: java-webflux/src/main/java/com/ederfmatos/javawebflux/entity/Transaction.java
package com.ederfmatos.javawebflux.entity;

import com.ederfmatos.javawebflux.enums.Currency;
import java.math.BigDecimal;
import java.time.ZonedDateTime;
import java.util.UUID;

public record Transaction(
        String id,
        String description,
        BigDecimal amount,
        BigDecimal totalAmount,
        Currency currency,
        ZonedDateTime createdAt
) {
    public static Transaction create(String description, BigDecimal amount, Currency currency, BigDecimal totalAmount) {
        return new Transaction(UUID.randomUUID().toString(), description, amount, totalAmount, currency, ZonedDateTime.now());
    }
}

File: kotlin-coroutines/src/main/kotlin/com/ederfmatos/kotlincoroutines/entity/Transaction.kt
package com.ederfmatos.kotlincoroutines.entity

import com.ederfmatos.kotlincoroutines.enums.Currency
import java.math.BigDecimal
import java.time.ZonedDateTime
import java.util.UUID

data class Transaction(
    val id: String = UUID.randomUUID().toString(),
    val description: String,
    val currency: Currency,
    val amount: BigDecimal,
    val totalAmount: BigDecimal,
    val createdAt: ZonedDateTime
) {

    companion object {
        fun create(description: String, totalAmount: BigDecimal, amount: BigDecimal, currency: Currency): Transaction {
            return Transaction(
                description = description,
                currency = currency,
                amount = amount,
                totalAmount = totalAmount,
                createdAt = ZonedDateTime.now()
            )
        }
    }
}